File: app/src/main/java/br/odb/angstronme/PlayGameActivity.java
package br.odb.angstronme;

import android.app.Activity;
import android.os.Bundle;
import android.view.KeyEvent;

public class PlayGameActivity extends Activity {


    private static final String DEBUG_TAG = "WW3D";
    Angtron3DGameCanvas canvas;

    Runnable updateRunnable = new Runnable() {
        public void run() {
            canvas.invalidate();
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_play_game);

        canvas = (Angtron3DGameCanvas) findViewById(R.id.gameView);
        canvas.RestartGame(getIntent().getIntExtra( "level", 0 ) * 5 );
    }

    void updateOnMainThread() {
        runOnUiThread(updateRunnable);
    }

    @Override
    protected void onResume() {
        super.onResume();

        new Thread(new Runnable() {
            public void run() {
                while (true) {
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    updateOnMainThread();
                }
            }
        }).start();

        new Thread(new Runnable() {
            public void run() {
                while (true) {
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    canvas.update();
                }
            }
        }).start();

        canvas.requestFocus();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {

        if ( keyCode == KeyEvent.KEYCODE_BACK ) {
            finish();
            return true;
        }

        return canvas.onKeyDown(keyCode, event);
    }
}

File: app/src/main/java/br/odb/angstronme/PlayGameActivity.kt
package br.odb.angstronme

import android.os.Bundle
import android.view.KeyEvent
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_play_game.*

class PlayGameActivity : AppCompatActivity() {
    private var updateRunnable = Runnable { gameView.invalidate() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_play_game)
        gameView.RestartGame(intent.getIntExtra("level", 0) * 5)
    }

    private fun updateOnMainThread() {
        runOnUiThread(updateRunnable)
    }

    override fun onResume() {
        super.onResume()
        Thread {
            while (true) {
                try {
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
                updateOnMainThread()
            }
        }.start()
        Thread {
            while (true) {
                try {
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
                gameView.update()
            }
        }.start()
        gameView.requestFocus()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            finish()
            return true
        }
        return gameView.onKeyDown(keyCode, event)
    }

    companion object {
    }
}